mavlink_vehicles/mavlink_vehicles.cc

```diff
@@ -985,12 +988,22 @@ void mav_vehicle::takeoff()
 {
     switch (this->flight_stack_type) {
     case firmware_type::PX4:
-        this->set_mode(mode::TAKEOFF);
+        if (this->get_status() != status::ACTIVE) {
+            this->set_mode(mode::TAKEOFF);
+        }
         break;
     case firmware_type::APM:
-        send_cmd_long(MAV_CMD_NAV_TAKEOFF, 0, 0, 0, 0, 0, 0,
-                      defaults::takeoff_init_alt_m,
-                      request_intervals_ms::takeoff);
+        if (this->get_mode() != mode::GUIDED) {
```

Comment:
Previously, the user of the API would have to execute all these steps before sending a takeoff command to the APM, what is not necessary for PX4.
Since we want to have APM and PX4 interchangeable transparently, I've added those commands to takeoff(). That way, the user would simply need to call takeoff() for both flight stacks. I'll submit a patch to gazebo-sitl with that change as well.